File: Android/ProjectoESeminario/app/src/main/java/andre/pt/projectoeseminario/Controller/State/ClipboardController.java
package andre.pt.projectoeseminario.Controller.State;

import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

/**
 * Helper so we can now when we copied new content.
 * This class is thread-safe
 */
public class ClipboardController {

    private Lock nLock;
    private final List<Pair> sentRequestsQueue;
    private final List<String> ignoreQueue;
    private long orderNumber = -1;

    //Helper class
    public class Pair {
        private String value;
        private boolean wake;
        private boolean remove;
        private Condition condition;

        public Pair(String value, Condition condition) {
            this.value = value;
            this.condition = condition;
            this.wake = false;
            this.remove = false;
        }

        public String getValue() {
            return value;
        }

        public Condition getCondition() {
            return condition;
        }

        public void Wake() {
            this.wake = true;
        }

        public void Remove() {
            this.remove = true;
        }

        public boolean isWake() {
            return wake;
        }

        public boolean toRemove() {
            return remove;
        }
    }

    public ClipboardController(long order){
        this.nLock = new ReentrantLock();
        this.sentRequestsQueue = new LinkedList<>();
        this.ignoreQueue = new LinkedList<>();

    }

    /**
     * Notifies the end of the putValue request.
     */
    public void wake(){
        nLock.lock();

        try {
            if(sentRequestsQueue.size() > 0)
                sentRequestsQueue.get(0).getCondition().signal();
        }finally {
            nLock.unlock();
        }

    }

    /**
     * Changes to the new value, if it is different that the old value
     * @param value the copied text
     * @return boolean indicating weather we changed value or not.
     */
    public boolean putValue(String value, Consumer<Pair> startTimer) {
        nLock.lock();

        try {
            if (ignoreQueue.contains(value))
            {
                ignoreQueue.remove(value);
                return false;
            }

            Pair node = new Pair(value, nLock.newCondition());
            sentRequestsQueue.add(node);
            startTimer.accept(node);

            do
            {
                try
                {
                    node.getCondition().await();
                }
                catch (InterruptedException e) {}

                if (node.toRemove())
                {
                    sentRequestsQueue.remove(node);
                    return false;
                }

                if (node == sentRequestsQueue.get(0) && node.isWake() && !node.toRemove())
                {
                    sentRequestsQueue.remove(node);

                    if (ignoreQueue.contains(value))
                    {
                        ignoreQueue.remove(value);
                        return false;
                    }

                    return true;
                }

            } while (true);
        }finally {
            nLock.unlock();
        }

    }


    //return 0 -> Didnt change
    //return 1 -> Did change, we need to update history
    //return 2 -> Old request, we need to update history
    public int PutValue(String text, long order_received)
    {
        nLock.lock();

        try {
            //When we dont have any pending request
            if (sentRequestsQueue.size() == 0)
                return AvaliateAndUpdateOrder(order_received, orderNumber, text);

            //Even though we didnt finish the requests we can be sure that the request received is more recent
            // Than our sent requests.
            return AvaliateAndUpdateOrder(order_received, orderNumber + sentRequestsQueue.size(), text);

        }finally {
            nLock.unlock();
        }
    }

    private int AvaliateAndUpdateOrder(long received, long actual, String text)
    {
        if (received == actual)
            return 0;

        if (received < actual)
            return 2;

        orderNumber = received;
        ignoreQueue.add(text);
        return 1;
    }

    public void updateStateOfUpload(long order)
    {
        nLock.lock();

        try {
            if(orderNumber < order)
                orderNumber = order;

            if (sentRequestsQueue.size() > 0)
            {
                Pair node = sentRequestsQueue.get(0);
                ignoreQueue.add(node.getValue());
                node.Wake();
                node.getCondition().signal();
            }

        }finally {
            nLock.unlock();
        }
    }

    public void removeUpload(Pair node)
    {
        nLock.lock();

        try {
            ignoreQueue.add(node.getValue());
            node.Remove();
            node.condition.signal();

        }finally {
            nLock.unlock();
        }
    }

    public void addToIgnoreQueue(String text)
    {
        nLock.lock();

        try{
            ignoreQueue.add(text);
        }finally {
            nLock.unlock();
        }
    }

    public long getOrder(){
        return orderNumber;
    }
}

File: Windows/ProjectoESeminario/ProjectoESeminario/Controller/State/ClipboardController.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace ProjectoESeminario.Controller.State
{
    public class ClipboardController
    {
        private readonly Object nLock;
        private long order_number;

        //Used for the sync problems discuted on the paper.Use pair object to facilitate the cancelation
        private readonly LinkedList<Pair> sentRequestsQueue;
        private readonly LinkedList<String> ignoreQueue;


        //Helper class
        public class Pair
        {
            private bool wake;
            private bool remove;
            private string value;

            public Pair(string text)
            {
                this.wake = false;
                this.remove = false;
                this.value = text;
            }

            public string GetValue()
            {
                return value;
            }

            public bool isWake()
            {
                return wake;
            }

            public void SetRemove()
            {
                this.remove = true;
            }

            public bool ToRemove()
            {
                return remove;
            }

            public void Wake()
            {
                this.wake = true;
            }
        }

        public ClipboardController()
        {
            nLock = new Object();
            sentRequestsQueue = new LinkedList<Pair>();
            ignoreQueue = new LinkedList<String>();
        }

        public void Wake()
        {
            lock (nLock)
            {

                if (sentRequestsQueue.Count > 0)
                    MonitorEx.Pulse(nLock, sentRequestsQueue.First);
                  
            }
        }

        public bool PutValue(String text, Action<LinkedListNode<Pair>> startTimer)
        {
            lock (nLock)
            {
                if (ignoreQueue.Contains(text))
                {
                    ignoreQueue.Remove(text);
                    return false;
                }

                var node = sentRequestsQueue.AddLast(new Pair(text));
                startTimer.Invoke(node);
                long order = order_number;

                do
                {
                    try
                    {
                        MonitorEx.Wait(nLock, node);
                    }
                    catch (ThreadInterruptedException) {}

                    if (node.Value.ToRemove())
                    {
                        sentRequestsQueue.Remove(node);
                        return false;
                    }

                    if (node == sentRequestsQueue.First && node.Value.isWake())
                    {
                        sentRequestsQueue.Remove(node);
                        return order_number != order;
                    }

                } while (true);

            }
        }

        //return 0 -> Didnt change
        //return 1 -> Did change, we need to update history
        //return 2 -> Old request, we need to update history
        public int PutValue(string text, long order_received)
        {
            lock (nLock)
            {
                //When we dont have any pending request
                if (sentRequestsQueue.Count == 0)
                    return AvaliateAndUpdateOrder(order_received, order_number, text);

                //Even though we didnt finish the requests we can be sure that the request received is more recent
                // Than our sent requests.
                return AvaliateAndUpdateOrder(order_received, order_number + sentRequestsQueue.Count, text);

            }
        }

        private int AvaliateAndUpdateOrder(long received, long actual, string text)
        {
            if (received == actual)
                return 0;

            if (received < actual)
                return 2;

            order_number = received;
            ignoreQueue.AddLast(text);
            return 1;
        }

        public void UpdateStateOfUpload(long order)
        {
            lock (nLock)
            {
                if(order_number < order)
                    order_number = order;

                if (sentRequestsQueue.Count > 0)
                {
                    var node = sentRequestsQueue.First;
                    ignoreQueue.AddLast(node.Value.GetValue());
                    node.Value.Wake();
                    MonitorEx.Pulse(nLock, node);

                }
            }
        }

        public void RemoveUpload(LinkedListNode<Pair> node)
        {
            lock (nLock)
            {
                ignoreQueue.AddLast(node.Value.GetValue());
                node.Value.SetRemove();
                MonitorEx.Pulse(nLock, node);
            }
        }

        public void AddToIgnoreQueue(string text)
        {
            lock (nLock)
            {
                ignoreQueue.AddLast(text);
            }
        }
        
    }
}
